ingest-cyclonedx-sbom/src/index.ts

```diff
@@ -0,0 +1,245 @@
+  @typescript-eslint/no-use-before-define,
+  @typescript-eslint/no-explicit-any, 
+*/
+
+function usage(): void {
+  console.log('ingest-cyclonedx-sbom --sbom <path> --repo <string> [--devDeps <path>] [--directDeps <path>] [--directOnly]');
+  console.log('  --sbom  path to sbom file to ingest');
+  console.log('  --repo  name of coderepo ingest ');
+  console.log('  (optional) --devDeps  path to json file array of dev dependencies');
+  console.log('  (optional) --directDeps  path to json file array of direct dependencies');
+  console.log('  (optional) --directOnly  only ingest direct dependencies (requires --directDeps)');
+}
+
+async function run (
+  pathToSBOM: string, 
+  repoName: string,
+  devDependencies: string[],
+  directDependencies: string[]
+): Promise<void> {
+  const SBOMComponents = await getSBOMComponents(pathToSBOM);
+
+  const j1Client = await initJ1Client();
+
+  console.log('Gathering CodeRepo data from JupiterOne...');
+  const codeRepoIdMap = await getCodeRepoIdMap(j1Client);
+  console.log(`${Object.keys(codeRepoIdMap).length} coderepos found in J1`);
+
+  const codeRepoEntityId = codeRepoIdMap[repoName];
+  if (!codeRepoEntityId) {
+    die(`ERROR: ${repoName} doesn't appear to be a CodeRepo known to JupiterOne!`, 3);
+  }
+
+  console.log('Gathering CodeModule data from JupiterOne...');
+  const codeModuleIdMap = await getCodeModuleIdMap(j1Client);
+  console.log(`${Object.keys(codeModuleIdMap).length} CodeModules found in J1`);
+
+  let ingestedEntitiesCount = 0;
+  for (const component of SBOMComponents) {
+    let res: any;
+    const purl = getPurl(component);
+    let packageId = codeModuleIdMap[purl];
+
+    // normalize component name for lookup in optional hint arrays
+    const normalizedComponentName = component.group ? component.group + '/' + component.name : component.name;
+
+    if (directOnly && !directDependencies.includes(normalizedComponentName)) {
+      console.log(`Skipping transitive dep: ${normalizedComponentName}`);
+      continue; // skip transitive deps if --directOnly is specified
+    }
+
+    if (!packageId) {
+
+      const dependencyProps: J1Dependency = {
+        displayName: component.name,
+        fullName: purl,
+        name: component.name,
+        purl,
+        license: getLicense(component),
+        scope: component.group,
+        version: component.version,
+        from: `ingest-cyclonedx-sbom-${repoName}`
+      };
+
+      // e.g. pkg:npm/commander@8.3.0 ingests as type 'npm_package'
+      const j1Type = getType(component);
+
+      process.stdout.write(`Creating CodeModule entity for ${purl} ${j1Type} in J1..`);
+      res = await retry(() => {
+        return j1Client.createEntity(
+          `${j1Type}:${component.name}`, //_key
+          j1Type, //_type
+          [ 'CodeModule' ], //_class
+          dependencyProps
+        );
+      }, attemptOptions);
+      console.log(' ok');
+      packageId = res.vertex.entity._id;
+      ingestedEntitiesCount++;
```

Comment:
I presume this will always be incremented whether or not the network request is successful.